File: Graph/DFS/Course Schedule/CourseSchedule.java
// Cycle detection. DFS or Topological Sort. Nodes for courses, directed edges for prereq -> course. Cycle implies impossible completion: need Course A to finish Course B, also need Course B to finish Course A. Contradiction.
public class CourseSchedule {
	public boolean canFinish(int numCourses, int[][] prerequisites) {
		Map<Integer, List<Integer>> graph = new HashMap<>();
		for (int[] prereq : prerequisites) {
			graph.computeIfAbsent(prereq[1], k -> new ArrayList<>()).add(prereq[0]);
		}
		int[] visited = new int[numCourses];
		for (int course = 0; course < numCourses; course++) {
			if (dfsHasCycle(course, graph, visited)) return false;
		}
		return true;
	}
	// white grey black
	public boolean dfsHasCycle(int course, Map<Integer, List<Integer>> graph, int[] visited) {
		if (visited[course] == 1) return true;
		if (visited[course] == 2) return false;
		
		visited[course] = 1; // mark visiting
		for (int neighbor : graph.getOrDefault(course, new ArrayList<>())) {
			if (dfsHasCycle(neighbor, graph, visited)) return true;
		}
		visited[course] = 2; // mark visited
		return false;
	}
}

File: Graph/DFS/Course Schedule/course_schedule.cpp
class CourseSchedule {
public:
    // Create a graph where c1 -> c2 denotes c1 is a prereq for c2. A cycle in this graph <=> impossible to complete courses. Kahn's Topological or white grey black DFS. 
    // Runtime: building graph -> O(V + E), DFS -> O(V + E) --> O(V + E) 
    // Space: adjacency list -> O(V + E) --> O(V + E)
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        // build graph
        unordered_map<int, vector<int>> graph;
        for (const auto& prereq : prerequisites) {
            graph[prereq[1]].push_back(prereq[0]);
        }        

        vector<int> visited(numCourses, 0);
        for (int course = 0; course < numCourses; ++course) {
            if (hasCycle(graph, course, visited)) return false;
        }

        return true;
    }

private:
    ;
    bool hasCycle(unordered_map<int, vector<int>>& graph, int course, vector<int>& visited) {
        if (visited[course] == 2) return false; // visisted
        if (visited[course] == 1) return true; // visiting

        visited[course] = 1;
        for (int neighbor : graph[course]) {
            if (hasCycle(graph, neighbor, visited)) return true;
        }
        visited[course] = 2;
        return false;
    }

};
